File: native/tests/PolicyProbe.swift
import Foundation

@main
enum PolicyProbe {
    private static func fail(_ message: String) -> Never {
        FileHandle.standardError.write(Data("\(message)\n".utf8))
        exit(2)
    }

    static func main() {
        let arguments = CommandLine.arguments
        guard arguments.count >= 3 else {
            fail("usage: PolicyProbe <operation> <configured-origin> [operation-inputs]")
        }

        let operation = arguments[1]
        let configuredOrigin = arguments[2]

        switch operation {
        case "configuration":
            guard arguments.count == 3 else { fail("invalid configuration arguments") }
            print(StoreDestination(origin: configuredOrigin)?.origin ?? "INVALID")
        case "navigation":
            guard arguments.count == 4 else { fail("invalid navigation arguments") }
            print(StoreDestination(origin: configuredOrigin)?.allows(arguments[3]) ?? false)
        case "bootstrap":
            guard arguments.count == 7,
                  let status = Int(arguments[3]),
                  arguments[5] == "true" || arguments[5] == "false" else {
                fail("invalid bootstrap arguments")
            }
            let distribution = arguments[6] == "NULL" ? nil : arguments[6]
            let accepted = StoreDestination(origin: configuredOrigin)?.acceptsBootstrap(
                status: status,
                finalURL: arguments[4],
                redirected: arguments[5] == "true",
                distribution: distribution
            ) ?? false
            print(accepted)
        default:
            fail("unknown operation")
        }
    }
}

File: native/tests/PolicyProbe.java
import pro.thesquad.shell.StoreDestination;

public final class PolicyProbe {
    private static void fail(String message) {
        System.err.println(message);
        System.exit(2);
    }

    public static void main(String[] arguments) {
        if (arguments.length < 2) {
            fail("usage: PolicyProbe <operation> <configured-origin> [operation-inputs]");
        }

        String operation = arguments[0];
        String configuredOrigin = arguments[1];

        switch (operation) {
            case "configuration" -> {
                if (arguments.length != 2) {
                    fail("invalid configuration arguments");
                }
                StoreDestination destination = StoreDestination.parse(configuredOrigin);
                System.out.println(destination == null ? "INVALID" : destination.origin());
            }
            case "navigation" -> {
                if (arguments.length != 3) {
                    fail("invalid navigation arguments");
                }
                StoreDestination destination = StoreDestination.parse(configuredOrigin);
                System.out.println(destination != null && destination.allows(arguments[2]));
            }
            case "bootstrap" -> {
                if (arguments.length != 6) {
                    fail("invalid bootstrap arguments");
                }
                int status;
                try {
                    status = Integer.parseInt(arguments[2]);
                } catch (NumberFormatException exception) {
                    fail("invalid bootstrap status");
                    return;
                }
                if (!arguments[4].equals("true") && !arguments[4].equals("false")) {
                    fail("invalid redirected value");
                }
                String distribution = arguments[5].equals("NULL") ? null : arguments[5];
                StoreDestination destination = StoreDestination.parse(configuredOrigin);
                boolean accepted = destination != null && destination.acceptsBootstrap(
                    status,
                    arguments[3],
                    Boolean.parseBoolean(arguments[4]),
                    distribution
                );
                System.out.println(accepted);
            }
            default -> fail("unknown operation");
        }
    }
}
